src/Razor/test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/Shared/DocumentExcerptServiceTestBase.cs

```diff
@@ -0,0 +1,117 @@
+        private (DocumentSnapshot primary, Document secondary) InitializeDocument(SourceText sourceText)
+        {
+            var project = new DefaultProjectSnapshot(
+                ProjectState.Create(Workspace.Services, HostProject)
+                .WithAddedHostDocument(HostDocument, () =>
+                {
+                    return Task.FromResult(TextAndVersion.Create(sourceText, VersionStamp.Create()));
+                }));
+
+            var primary = project.GetDocument(HostDocument.FilePath);
+
+            var solution = Workspace.CurrentSolution.AddProject(ProjectInfo.Create(
+                ProjectId.CreateNewId(Path.GetFileNameWithoutExtension(HostDocument.FilePath)),
+                VersionStamp.Create(),
+                Path.GetFileNameWithoutExtension(HostDocument.FilePath),
+                Path.GetFileNameWithoutExtension(HostDocument.FilePath),
+                LanguageNames.CSharp,
+                HostDocument.FilePath));
+
+            solution = solution.AddDocument(
+                DocumentId.CreateNewId(solution.ProjectIds.Single(), HostDocument.FilePath),
+                HostDocument.FilePath,
+                new GeneratedDocumentTextLoader(primary, HostDocument.FilePath));
+
+            var secondary = solution.Projects.Single().Documents.Single();
+            return (primary, secondary);
+        }
+
+        // Maps a span in the primary buffer to the secondary buffer. This is only valid for C# code
+        // that appears in the primary buffer.
+        private async Task<TextSpan> GetSecondarySpanAsync(DocumentSnapshot primary, TextSpan primarySpan, Document secondary)
+        {
+            var output = await primary.GetGeneratedOutputAsync();
+
+            var mappings = output.GetCSharpDocument().SourceMappings;
+            for (var i = 0; i < mappings.Count; i++)
+            {
+                var mapping = mappings[i];
+                if (mapping.OriginalSpan.AbsoluteIndex <= primarySpan.Start &&
+                    (mapping.OriginalSpan.AbsoluteIndex + mapping.OriginalSpan.Length) >= primarySpan.End)
+                {
+                    var offset = mapping.GeneratedSpan.AbsoluteIndex - mapping.OriginalSpan.AbsoluteIndex;
+                    var secondarySpan = new TextSpan(primarySpan.Start + offset, primarySpan.Length);
+                    Assert.Equal(
+                        (await primary.GetTextAsync()).GetSubText(primarySpan).ToString(),
+                        (await secondary.GetTextAsync()).GetSubText(secondarySpan).ToString());
+                    return secondarySpan;
+                }
+            }
+
+            throw new InvalidOperationException("Could not map the primary span to the generated code.");
+        }
+
+        public async Task<(Document generatedDocument, SourceText razorSourceText, TextSpan primarySpan, TextSpan generatedSpan)> InitializeAsync(string razorSource)
```

Comment:
Two separate initializers to deal with access restrictions. As this was only for testing purposes, figured it would be acceptable.